plugins/micron/micron-utils.c

```diff
@@ -0,0 +1,286 @@
+
+				text[0] = dec;
+			}
+		}
+
+		text[symbol_count] = '\0'; // Terminate the string
+	}
+
+	return text;
+}
+
+unsigned char *read_binary_file(char *data_dir_path, const char *bin_path,
+								long *buffer_size, int retry_count)
+{
+	char *file_path = NULL;
+	FILE *bin_file = NULL;
+	size_t n_data = 0;
+	unsigned char *buffer = NULL;
+
+	/* set path */
+	if (data_dir_path == NULL) {
+		file_path = (char *)bin_path;
+	} else {
+		/* +2 for the / and null terminator */
+		file_path = (char *) calloc(1, strlen(data_dir_path) + strlen(bin_path) + 2);
+		if (!file_path)
+			return NULL;
+
+		if (strlen(bin_path) != 0)
+			sprintf(file_path, "%s/%s", data_dir_path, bin_path);
+		else
+			sprintf(file_path, "%s", data_dir_path);
+	}
+
+	/* open file */
+	for (int i = 0; i < retry_count; i++) {
+		bin_file = fopen(file_path, "rb");
+		if (bin_file != NULL)
+			break;
+		sleep((unsigned int)(retry_count > 1));
+	}
+
+	if (!bin_file) {
+		nvme_show_error("\nFailed to open %s", file_path);
+		if (file_path != bin_path)
+			free(file_path);
+		return NULL;
+	}
+
+	/* get size */
+	fseek(bin_file, 0, SEEK_END);
+	*buffer_size = ftell(bin_file);
+	fseek(bin_file, 0, SEEK_SET);
+	if (*buffer_size <= 0) {
+		fclose(bin_file);
+		return NULL;
+	}
+
+	/* allocate buffer */
+	buffer = (unsigned char *)malloc(*buffer_size);
+	if (!buffer) {
+		nvme_show_result("\nFailed to allocate %ld bytes!", *buffer_size);
+		fclose(bin_file);
+		return NULL;
+	}
+	memset(buffer, 0, *buffer_size);
+
+	/* Read data */
+	n_data = fread(buffer, 1, *buffer_size, bin_file);
+
+	/* Close file */
+	fclose(bin_file);
+
+	/* Validate we read data */
+	if (n_data != (size_t)*buffer_size) {
+		nvme_show_result("\nFailed to read %ld bytes from %s", *buffer_size, file_path);
+		return NULL;
+	}
+
+	if (file_path != bin_path)
+		free(file_path);
+	return buffer;
+}
+
+void print_formatted_var_size_str(const char *msg, const __u8 *pdata, size_t data_size, FILE *fp)
+{
+	char description_str[256] = "";
+	char temp_buffer[3] = { 0 };
+
+	for (size_t i = 0; i < data_size; ++i) {
+		sprintf(temp_buffer, "%02X", pdata[i]);
+		strcat(description_str, temp_buffer);
+	}
+
+	if (fp)
+		fprintf(fp, "%s: %s\n", msg, description_str);
+	else
+		printf("%s: %s\n", msg, description_str);
+}
+
+void print_micron_vs_logs(__u8 *buf, struct micron_vs_logpage *log_page, int field_count,
+				 struct json_object *stats, __u8 spec, FILE *fp)
+{
+	__u64 lval_lo, lval_hi;
+	__u32 ival;
+	__u16 sval;
+	__u8  cval, lval[8] = { 0 };
+	int field;
+	int offset = 0;
+
+	for (field = 0; field < field_count; field++) {
+		char datastr[1024] = { 0 };
+		char *sfield = log_page[field].field;
+		int size = !spec ? log_page[field].size : log_page[field].size2;
+
+		if (!size || sfield == NULL)
+			continue;
+
+		switch (size) {
+		case FIELD_SIZE_16: {
+			if (strstr(sfield, "GUID")) {
+				sprintf(datastr, "0x%"PRIx64"%"PRIx64"",
+						(uint64_t)le64_to_cpu(*(uint64_t *)(&buf[offset + 8])),
```

Comment:
Is this casting really correct since seems below correct?
```diff
- (uint64_t)le64_to_cpu(*(uint64_t *)(&buf[offset + 8])),
+ le64_to_cpu(*(__u64 *)(&buf[offset + 8])),
```

Reply:
corrected, thanks.